pyrefly/lib/alt/class/dataclass.rs

```diff
@@ -137,17 +144,39 @@ impl<'a, Ans: LookupAnswer> AnswersSolver<'a, Ans> {
         cls: &Class,
         fields: &SmallSet<Name>,
         kw_only: bool,
+        errors: &ErrorCollector,
     ) -> ClassSynthesizedField {
         let mut params = vec![self.class_self_param(cls, false)];
+        let mut has_seen_default = false;
+
         for (name, field, field_flags) in self.iter_fields(cls, fields, true) {
             if field_flags.is_set(&DataclassKeywords::INIT) {
-                params.push(field.as_param(
-                    &name,
-                    field_flags.is_set(&DataclassKeywords::DEFAULT),
-                    field_flags.is_set(&(DataclassKeywords::KW_ONLY.0, kw_only)),
-                ));
+                let has_default = field_flags.is_set(&DataclassKeywords::DEFAULT);
+                let is_kw_only = field_flags.is_set(&(DataclassKeywords::KW_ONLY.0, kw_only));
+
+                if !is_kw_only {
+                    if !has_default && has_seen_default {
+                        if let Some(range) = cls.field_decl_range(&name) {
+                            self.error(
+                                errors,
+                                range,
+                                error::kind::ErrorKind::BadClassDefinition,
+                                None,
+                                format!(
+                                    "DataClass field '{}' without a default may not follow DataClass field with a default", name
```

Comment:
nit: dataclass is usually written all lower-case.
```suggestion
                                    "Dataclass field '{}' without a default may not follow dataclass field with a default", name
```